python/lsst/daf/butler/_butler_factory.py

```diff
@@ -0,0 +1,165 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("ButlerFactory",)
+
+from typing import Callable, Mapping
+
+from lsst.resources import ResourcePathExpression
+
+from ._butler import Butler
+from ._butler_config import ButlerConfig, ButlerType
+from ._butler_repo_index import ButlerRepoIndex
+from ._utilities.named_locks import NamedLocks
+from ._utilities.thread_safe_cache import ThreadSafeCache
+
+_FactoryFunction = Callable[[str | None], Butler]
+"""Function that takes an access token string or `None`, and returns a Butler
+instance."""
+
+
+class ButlerFactory:
+    """Factory for efficiently instantiating Butler instances from the
```

Comment:
Now that I'm reading #931, we seem to be saying that server is required to use a label from a DAF_BUTLER_REPOSITORY_INDEX.  This makes me ponder that the name of the class is a bit too generic since DirectButler is not required to use the index labels and we don't expect this class to be used in non-server context.

Reply:
Well, it's DAF_BUTLER_REPOSITORY_INDEX or "pass in whatever set of labels and config files you want."  I'd be open to renaming if you have any suggestions.

Reply:
Since it has to be able to return DirectButler and RemoteButler for now the class can't really be ServerButlerFactory (as is the name of the variable you assign to). `LabeledButlerFactory`? `NamedButlerFactory`?